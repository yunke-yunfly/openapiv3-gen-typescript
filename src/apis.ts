import type { OpenAPIV3 } from 'openapi-types';

import type { Api, ResolverType } from './resolver';

// 获取简化的数据
export function getApis(openApiDoc: OpenAPIV3.Document): Api[] {
  const paths = openApiDoc.paths;
  return Object.keys(paths)
    .map((url) => {
      const methods: string[] = Object.keys(paths[url] || {});
      return methods.map((method) => {
        const {
          description,
          operationId,
          summary,
          tags,
          parameters = [],
          requestBody,
          responses,
        } = ((paths[url] as any)?.[method] || {}) as OpenAPIV3.OperationObject;
        const parametersObject = parameters.filter(
          (item) => !(item as OpenAPIV3.ReferenceObject).$ref,
        ) as OpenAPIV3.ParameterObject[];
        return {
          url,
          method,
          description,
          operationId,
          summary,
          tags,
          params: parametersObject.filter((item) => item.in === 'path'),
          query: parametersObject.filter((item) => item.in === 'query'),
          cookie: parametersObject.filter((item) => item.in === 'cookie'),
          header: parametersObject.filter((item) => item.in === 'header'),
          response: Object.keys(responses).reduce((acc, statusCode) => {
            if (!(responses[statusCode] as OpenAPIV3.ReferenceObject)['$ref']) {
              const { content = {}, description } = responses[
                statusCode
              ] as OpenAPIV3.ResponseObject;
              const schema = Object.keys(content).map((type) => content[type].schema);

              acc[statusCode] = {
                description,
                schema: schema[0],
              };
            } else {
              acc[statusCode] = {
                schema: responses[statusCode] as OpenAPIV3.ReferenceObject,
              };
            }
            return acc;
          }, {} as Record<string, { description?: string; schema?: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject }>),
          body: (() => {
            if (!requestBody) return;
            return (requestBody as OpenAPIV3.ReferenceObject)?.['$ref']
              ? (requestBody as OpenAPIV3.ReferenceObject)
              : Object.keys((requestBody as OpenAPIV3.RequestBodyObject)?.content).map(
                  (type) => (requestBody as OpenAPIV3.RequestBodyObject)?.content[type].schema,
                )[0];
          })(),
        };
      });
    })
    .flat();
}

// 解析数据
export type ResolvedApisType = Record<keyof ResolverType, string | undefined>;
export function resolveApis(
  apis: Api[],
  resolver: Required<ResolverType>,
): Promise<ResolvedApisType[]> {
  return Promise.all(
    apis.map(async (api) => ({
      fileName: resolver.fileName(api),
      url: resolver.url(api),
      method: resolver.method(api),
      fnComment: resolver.fnComment(api),
      fnName: resolver.fnName(api),
      params: await resolver.params(api),
      query: await resolver.query(api),
      cookie: await resolver.cookie(api),
      header: await resolver.header(api),
      body: await resolver.body(api),
      response: await resolver.response(api),
    })),
  );
}
