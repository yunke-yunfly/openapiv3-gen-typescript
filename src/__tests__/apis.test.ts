import { getApis } from '../apis';

describe('apis', () => {
  test('getApis', () => {
    const openapi = require('./resource/openapi.json');
    expect(getApis(openapi)).toMatchSnapshot();
  });
});
