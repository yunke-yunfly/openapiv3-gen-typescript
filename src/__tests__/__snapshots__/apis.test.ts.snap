// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apis getApis 1`] = `
Array [
  Object {
    "body": undefined,
    "cookie": Array [],
    "description": "B_ActivityController.getActDetails  ",
    "header": Array [],
    "method": "get",
    "operationId": "B_ActivityController.get.getActDetails.bjWHnbNHPc",
    "params": Array [],
    "query": Array [
      Object {
        "description": "活动id (@QueryParam, 类型：string)",
        "in": "query",
        "name": "id",
        "required": true,
        "schema": Object {
          "type": "string",
        },
      },
      Object {
        "description": "活动场景值 (@QueryParam, 类型：number)",
        "example": "0",
        "in": "query",
        "name": "scene_type",
        "required": true,
        "schema": Object {
          "type": "number",
        },
      },
    ],
    "response": Object {
      "200": Object {
        "description": "",
        "schema": Object {
          "additionalProperties": false,
          "properties": Object {
            "code": Object {
              "type": "number",
            },
            "data": Object {
              "$ref": "#/components/schemas/GetActDetailsResponse",
            },
            "msg": Object {
              "type": "string",
            },
          },
          "required": Array [
            "code",
          ],
          "type": "object",
        },
      },
    },
    "summary": "后台获取活动详情",
    "tags": Array [
      "B_ActivityController",
    ],
    "url": "/tapi/b_activity/get_act_details",
  },
  Object {
    "body": Object {
      "properties": Object {
        "name": Object {
          "description": "@BodyParam",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
    "cookie": Array [],
    "description": "ExampleController.jest 测试案例controller ",
    "header": Array [],
    "method": "post",
    "operationId": "ExampleController.post.jest.GzbncJxNbQ",
    "params": Array [],
    "query": Array [],
    "response": Object {
      "200": Object {
        "description": "",
        "schema": Object {
          "properties": Object {
            "code": Object {
              "description": "接口返回code码字段",
              "type": "number",
            },
            "data": Object {
              "type": "string",
            },
            "msg": Object {
              "description": "接口返回信息字段",
              "type": "string",
            },
          },
          "required": Array [
            "code",
            "data",
          ],
          "type": "object",
        },
      },
    },
    "summary": "测试用例",
    "tags": Array [
      "ExampleController",
    ],
    "url": "/tapi/b_activity/get_act_details",
  },
]
`;
